src/test/kotlin/cacophony/semantic/analysis/VarUseTypeAnalysisTest.kt

```diff
@@ -98,21 +112,218 @@ class VarUseTypeAnalysisTest {
                     varUse to declaration,
                 ),
                 mapOf(programFunctionAnalysis(ast)),
+                variablesMap,
             )
         assertThat(result).containsAllEntriesOf(
             mapOf(
-                block to mapOf(declaration to VariableUseType.READ),
+                block to mapOf(variable to VariableUseType.READ),
             ),
         )
     }
 
+    @Test
+    fun `should find struct variables`() {
+        //       s
+        //      / \
+        //     /   \
+        //    a     b
+        //    |    / \
+        //    c   d   e
+        //        |
+        //        f
+        //
+        //
+        // f => ( let s, s.a = s.b.d)
+        val sDeclaration = variableDeclaration("s", Empty(mockRange()))
+        val aDeclaration = variableDeclaration("a", Empty(mockRange()))
+        val bDeclaration = variableDeclaration("b", Empty(mockRange()))
+        val cDeclaration = variableDeclaration("c", Empty(mockRange()))
+        val dDeclaration = variableDeclaration("d", Empty(mockRange()))
+        val eDeclaration = variableDeclaration("e", Empty(mockRange()))
+        val fDeclaration = variableDeclaration("f", Empty(mockRange()))
+        val fVariable = Variable.PrimitiveVariable()
+        val eVariable = Variable.PrimitiveVariable()
+        val dVariable = Variable.StructVariable(mapOf("f" to fVariable))
+        val cVariable = Variable.PrimitiveVariable()
+        val bVariable = Variable.StructVariable(mapOf("d" to dVariable, "e" to eVariable))
+        val aVariable = Variable.StructVariable(mapOf("c" to cVariable))
+        val sVariable = Variable.StructVariable(mapOf("a" to aVariable, "b" to bVariable))
+
+        val varSUse1 = variableUse("s")
+        val varSUse2 = variableUse("s")
+
+        val saLValue =
+            FieldRef.LValue(
+                mockRange(),
+                varSUse1,
+                "a",
+            )
+        val sbLValue =
+            FieldRef.LValue(
+                mockRange(),
+                varSUse2,
+                "b",
+            )
+        val sbdLValue =
+            FieldRef.LValue(
+                mockRange(),
+                sbLValue,
+                "d",
+            )
+
+        val assignment =
+            OperatorBinary.Assignment(
+                mockRange(),
+                saLValue,
+                sbdLValue,
+            )
+        val funF =
+            unitFunctionDefinition(
+                "f",
+                block(
+                    sDeclaration,
+                    assignment,
+                ),
+            )
+        val ast = astOf(funF)
```

Comment:
wrapping in function is not necessary here, we could do just astOf(declaration, assignment)